components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Runtime/Extensions/SecurityConfigEx.cs

```diff
@@ -45,18 +48,28 @@ public static class SecurityConfigEx {
                     $"{nameof(securityConfig)}.{nameof(SecurityConfig.ApplicationCertificate)}");
             }
 
-            var securityConfiguration = new SecurityConfiguration {
-                TrustedIssuerCertificates = securityConfig.TrustedIssuerCertificates.ToCertificateTrustList(),
-                TrustedPeerCertificates = securityConfig.TrustedPeerCertificates.ToCertificateTrustList(),
-                RejectedCertificateStore = securityConfig.RejectedCertificateStore.ToCertificateTrustList(),
-                AutoAcceptUntrustedCertificates = securityConfig.AutoAcceptUntrustedCertificates,
-                RejectSHA1SignedCertificates = securityConfig.RejectSha1SignedCertificates,
-                MinimumCertificateKeySize = securityConfig.MinimumCertificateKeySize,
-                ApplicationCertificate = securityConfig.ApplicationCertificate.ToCertificateIdentifier(hostname),
-                AddAppCertToTrustedStore = securityConfig.AddAppCertToTrustedStore
-            };
-
-            return securityConfiguration;
+            var options = applicationConfigurationBuilder
+                .AddSecurityConfiguration(
+                    securityConfig.ApplicationCertificate.SubjectName.Replace("localhost", hostname),
+                    securityConfig.PkiRootPath)
+                .SetAutoAcceptUntrustedCertificates(securityConfig.AutoAcceptUntrustedCertificates)
+                .SetRejectSHA1SignedCertificates(securityConfig.RejectSha1SignedCertificates)
+                .SetMinimumCertificateKeySize(securityConfig.MinimumCertificateKeySize)
+                .SetAddAppCertToTrustedStore(securityConfig.AddAppCertToTrustedStore);
+
+            securityConfig.ApplicationCertificate.BuildCertificateIdentifier(
+                applicationConfiguration.SecurityConfiguration.ApplicationCertificate);
```

Comment:
I find it quite confusing that here (`BuildCertificateTrustList()`, `BuildCertificateIdentifier()` and `BuildCertificateIdentifierStore()` methods) we are changing `applicationConfiguration` itself. I don't think that it is clear for the caller of those methods either, as well as the caller of `BuildSecurityConfiguration()` method itself that that is what will happen.

Since we are changing values of `applicationConfiguration`, it would have been much clearer if `BuildCertificateTrustList()` method returned a `CertificateTrustList` object that we would assign to `applicationConfiguration.SecurityConfiguration.TrustedPeerCertificates`. In its current form the whole logic is hidden, which makes it very hard for somebody else to understand later on.

Reply:
will be refactored soon